registration/api/v2/serializers.py

```diff
@@ -315,6 +344,88 @@ def validate_user(self, value):
         else:
             return value
 
+    def validate_external_domain(self, value) -> str:
+        """
+        Prevent users from registering with an external domain which was or currently in use.
+
+        The validation reduces the risk of security issues when someone is trying to take over
+        control of a client resource (domain) if they forget to restrict its access.
+        """
+        domain_data = tldextract.extract(value)
+        domain = domain_data.registered_domain
+
+        if domain == settings.DEFAULT_INSTANCE_BASE_DOMAIN:
+            raise ValidationError('The domain "{domain}" is not allowed.'.format(
+                domain=value
+            ))
+
+        self._validate_subdomain_not_contains_reserved_word(domain_data.subdomain)
+
+        is_domain_used_for_beta_app = BetaTestApplication.objects.filter(
+            external_domain__endswith=domain
+        ).exists()
+
+        if is_domain_used_for_beta_app:
+            raise ValidationError('This domain is already taken.')
+
+        is_taken = OpenEdXInstance.objects.filter(
+            Q(external_lms_domain=domain)
+            | Q(external_lms_preview_domain__endswith=domain)
+            | Q(external_studio_domain__endswith=domain)
+            | Q(external_discovery_domain__endswith=domain)
+            | Q(external_ecommerce_domain__endswith=domain)
+            # No need to check for subdomain, since it will match anyway
+            | Q(extra_custom_domains__contains=domain)
+        ).exists()
+
+        if is_taken:
+            raise ValidationError('This domain is already taken.')
+
+        return value
+
+    def validate_subdomain(self, value) -> str:
+        """
+        Prevent users from registering with a subdomain which is in use.
+
+        The validation reduces the risk of security issues when someone is trying to take over
+        control of a client resource (domain) if they forget to restrict its access.
+        """
+        truncated_subdomain = value.split(".")[0]
+        self._validate_subdomain_not_contains_reserved_word(truncated_subdomain)
+
+        # Although the validator on the model would catch the existence of
+        # the beta application with the given subdomain, this way we can potentially
+        # save an API request to Gandi for checking the existance of the DNS record.
+        is_subdomain_used_for_beta_app = BetaTestApplication.objects.filter(
+            subdomain=value
+        ).exists()
+
+        if is_subdomain_used_for_beta_app:
+            raise ValidationError("This domain is already taken.")
+
+        domains = set([
+            settings.DEFAULT_INSTANCE_BASE_DOMAIN,
+            settings.GANDI_DEFAULT_BASE_DOMAIN,
+        ])
+
+        for domain in domains:
+            try:
+                managed_domains = gandi_api.filter_dns_records(domain)
```

Comment:
This is going to return a _lot_ of records in production. Not sure we can do this. It may be easier and just as affective to have a blacklist of subdomains we definitely don't want used (eg. `*.net.opencraft.hosting` covers most cases) - expanding on the `_validate_subdomain_not_contains_reserved_word` method.

Reply:
I left a comment regarding this on the Jira task, but in general:

I don't feel that filtering over those records should be a big issue in the production environment considering the number of records, response time of the API and the ability to use a cache (if we want).

The reason I'd check the existing records anyway is that the effect on our DNS settings is much bigger if someone uses a not blacklisted record (which should be) than for the users to wait ~300 ms on the API request. I'm thinking about a scenario when a user is capable to overwrite for example a k8s cluster what we are building (in theory), because we forgot to list that in the blacklist.

I think it would be safer on the long run to check the records than facing with issues on the production env caused by a user.

Reply:
Yep, resolved on Jira. This is fine. :+1:

Reply:
@gabor-boros oh something I just thought of - is this serializer also used for updates? If so, is this then going to fail validation if the domain name wasn't changed, because that domain name will now be in the dns records?

Reply:
@swalladge Nice thought, I'll double check it 👍

Reply:
The final solution regarding serializers: The serializer will validate the field if and only if the field is changed compared to its previous state.

Reply:
Also fixed in 56747bf